docs/rfcs/0006-clustering.md

```diff
@@ -0,0 +1,331 @@
+  - Use an external store for configuration files
+  - Expose an API for managing configuration
+  - Running multiple configuration files at once.
+* Distribute load by merging multiple configuration files.
+
+## Proposal
+The proposal is based on prior art: https://github.com/grafana/agent/issues/872, https://github.com/grafana/agent/pull/1140
+
+* We will continue with a
+  [gossip-based](https://en.wikipedia.org/wiki/Gossip_protocol) approach using
+Hashicorp’s memberlist for our cluster
+* We will reuse the rfratto/ckit package code
+* We will use HTTP2 for communication between nodes
+* We will use go-discover for bootstrapping the cluster and discovering peers
+* A non-clustered Agent will work similar to a one-node cluster, which in the
+  future will be the default mode of operation
+
+## Implementation
+The feature will be behind a feature flag `--enable-features=clustering`. An
+agent can opt-in to clustering by passing a `--cluster.discover-peers`
+command-line flag with a comma-separated list of peers to connect to. Whenever
+an agent node receives a message about another node joining or leaving the
+cluster, it will propagate the message to its neighbors, and so on, until this
+information has reached all members of the cluster. The gossip memberlist will
+be utilized for the peer list, health checking and distribution of tokens
+between the agent nodes.
+
+All nodes will have access to a shared ckit.Sharder interface implementation
+which will be used to expose methods for each individual agent’s Flow
+controller to determine ownership of resources. As nodes enter and exit the
+cluster, the Sharder (eg. a consistent hashing ring) will redistribute tokens
+among the nodes in the cluster. The eventually consistent cluster state is when
+all nodes are working with the same configuration and have knowledge of each
+one of their peers in the cluster.The Sharder will be used to determine
+ownership of resources by hashing a desired value and checking the peers
+responsible for the corresponding ring token.
+
+When all nodes in the cluster have an up-to-date image of their peers, they
+will be able to independently agree to the ownership of a resource without
+having to communicate with each other, as local hashing will provide the same
+results for all nodes. For example, an agent node will be able to hash the
+label set of a target and check which of the peers is responsible for scraping
+that target. Similarly, an agent node will be able to hash the fully qualified
+component name and decide whether a component needs to be scheduled on this
+node or if another peer takes responsibility for it.
+
+On a more practical note, this clustering will most likely work with a
+[Kubernetes HPA](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/)
+so we can dynamically scale a fleet of agents. The target resource the
+autoscaler will look out for is most likely memory usage. In the future we may
```

Comment:
> most likely memory usage

I wonder what would happen if we have, say, 1 target with a huge number of data and 10 targets with very little data in their `/metrics` endpoint. In that case there'd be very little point in sharding because we cannot split this 1 big target in any way. I suppose we are making an assumption that the size of the `/metrics` payload is roughly the same for all targets, or at least that there are no big outliers? Not sure if it's worth documenting this.

Another use case could be having a much higher sampling frequency for some targets as opposed to others. The agents could be sharded in such a way that the more frequently sampled endpoints are given more resources.

Reply:
> I suppose we are making an assumption that the size of the /metrics payload is roughly the same for all targets, or at least that there are no big outliers?

Yes, your hunch is true; we're assuming that even if outliers exist, they are enough in numbers to be ~evenly distributed. The distribution will always be a little worse than hashmod sharding (as we're only distributing a certain number of tokens), but my back-of-the-envelope calculation in a realistic environment showed the stdev/mean over 4 nodes to be around 18% and not that worse than hashmod.

I've considered metrics-level sharding, but I think we shouldn't go about it right now. The target-level sharding is what we're already using with hashmod, is less resource intensive, easier to debug and seems works well enough for regular use cases.

> Another use case could be having a much higher sampling frequency for some targets as opposed to others. The agents could be sharded in such a way that the more frequently sampled endpoints are given more resources.

The scrape interval is dictated per `prometheus.scrape` component, so in this case, _all_ targets would be sampled at the same rate. Were you thinking about multiple scrape components running in parallel?

Reply:
> Were you thinking about multiple scrape components running in parallel?

Yes. If I understand correctly, in such case all scrape components will be sharded the same number of times. If we want to shard them differently, we would have to run two clusters. That sounds reasonable to me.